### ISSUE TITLE
Failure to install packages with "sudo: no tty present and no askpass program specified" error

### ISSUE THREAD

**Issue by mpinese** [JuliaLang/BinDeps.jl#199]
I've run into an odd problem with building packages with apt dependencies, in which the password prompt for the sudo apt-get fails.  At the moment a simple workaround is just installing the packages manually, but I'm posting what I've got in case you guys can see an easy fix.

The problem occurs on Lubuntu 15.10, when running julia 0.4.2 (staticfloat ppa release) from a local LXterminal session.  I've observed it on Cairo.jl ([_jl_libcairo error with Gadfly + Cairo #124](https://github.com/JuliaLang/Cairo.jl/issues/124#issuecomment-176607173) ) and HDF5.jl (see code block following).

```
julia> Pkg.build("HDF5")
INFO: Building Blosc
INFO: Building HDF5
Installing dependency hdf5-tools via `sudo apt-get install hdf5-tools`:
sudo: no tty present and no askpass program specified
================================[ ERROR: HDF5 ]=================================

LoadError: failed process: Process(`sudo apt-get install hdf5-tools`, ProcessExited(1)) [1]
while loading /home/mark/.julia/v0.4/HDF5/deps/build.jl, in expression starting on line 24

================================================================================
```

Hope that's of some use.  Let me know if you need any more logs etc. -- I know Lubuntu isn't a very common distro!  Thanks.

**Comment by zhenpan** [comment]
Hi Mark,

I got the same problem when Pkg.add("HDF5"). Could you please give more details about "how manually the packages" ?  Thank you.
Zhen

**Comment by yuyichao** [comment]
`sudo apt-get install hdf5-tools`

**Comment by zhenpan** [comment]
Yichao, thank you for the quick reply.

Here is the following problem:  I found HDF5 was actually installed along with my anaconda, so I insert the following command in  .juliarc.jl 

**push!(Libdl.LOAD_PATH, "/home/zhenpan/code/anaconda2/pkgs/hdf5-1.8.15.1-2/lib")**

then 

**Pkg.build("HDF5")**

signal (15): Terminated
while loading no file, in expression starting on line 0
syscall at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
uv__epoll_wait at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
uv__io_poll at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
uv_run at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
process_events at ./stream.jl:709
wait at ./task.jl:346
wait at ./task.jl:277
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
anonymous at multi.jl:603
unknown function (ip: 0x7f20e4ecb569)
unknown function (ip: (nil))
unknown function (ip: (nil))
ERROR: Base.Pkg.PkgError("HDF5 is not an installed package")
 [inlined code] from pkg.jl:13
 in build! at pkg/entry.jl:506
 in build! at pkg/entry.jl:551
 [inlined code] from dict.jl:473
 in build at pkg/entry.jl:573
 in anonymous at pkg/dir.jl:31
 in cd at file.jl:47
 in cd at pkg/dir.jl:31
 [inlined code] from pkg/dir.jl:25
 in build at pkg.jl:52
 in eval at ./boot.jl:263

Did I point julia to a wrong hdf5 lib path?

**Comment by zhenpan** [comment]
Yichao, thanks for the quick reply.

Here is the following problem. I found hdf5 was actually installed along with my anaconda, so I include the following command in .juliarc.jl

**push!(Libdl.DL_LOAD_PATH, "/home/zhenpan/code/anaconda2/pkgs/hdf5-1.8.15.1-2/lib")**

then 

**Pkg.build("HDF5")**

signal (15): Terminated
while loading no file, in expression starting on line 0
syscall at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
uv__epoll_wait at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
uv__io_poll at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
uv_run at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
process_events at ./stream.jl:709
wait at ./task.jl:346
wait at ./task.jl:277
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
anonymous at multi.jl:603
unknown function (ip: 0x7f61b581b569)
unknown function (ip: (nil))
unknown function (ip: (nil))
ERROR: Base.Pkg.PkgError("HDF5 is not an installed package")
 [inlined code] from pkg.jl:13
 in build! at pkg/entry.jl:506
 in build! at pkg/entry.jl:551
 [inlined code] from dict.jl:473
 in build at pkg/entry.jl:573
 in anonymous at pkg/dir.jl:31
 in cd at file.jl:47
 in cd at pkg/dir.jl:31
 [inlined code] from pkg/dir.jl:25
 in build at pkg.jl:52
 in eval at ./boot.jl:263

Did I point julia to a wrong lib path?

**Comment by yuyichao** [comment]
The error suggests that you havent add HDF5 yet

**Comment by zhenpan** [comment]
Thanks, Yichao.  Problem fixed. 

After 

sudo apt-get install hdf5-tools

and point julia to the system-wide hdf5 lib,  and 

Pkg.build("HDF5")

it works (though with another error in building Blosc, which was solved by Jingpeng [https://github.com/JuliaIO/HDF5.jl/issues/234] )

**Comment by nalimilan** [comment]
I'm seeing the same sudo error here on both 0.4.7 and master. Looks like it's not attached to the Julia terminal as it should, and therefore cannot ask for password.

**Comment by abelsiqueira** [comment]
It looks like the problem happen with `detach`, which is called by `Pkg.build`.
I can reproduce this problem with
```
run(detach(`sudo ls`))
```
`detach` calls `Cmd(..., detach=true)`, which is [here](https://github.com/JuliaLang/julia/blob/4e2e9312a06cf54e6ad90cd7232657d5faefacbe/base/process.jl#L21). After that I can't understand.

**Comment by ihnorton** [comment]
Your analysis looks correct. This is a likely unintended side-effect of the fix for https://github.com/JuliaLang/julia/issues/13458

We now have functionality to securely ask for passwords on all platforms (I think, [see here](https://github.com/JuliaLang/julia/pull/17506) and related links), so prompting could potentially be done by Julia. However, this would require some architectural changes, and Pkg is being redesigned right now anyway. For now the best option is probably to run installers manually before calling `Pkg.build`.

cc @StefanKarpinski -- not sure if this should go to the main Julia repo or the Pkg dev planner.

**Comment by nalimilan** [comment]
We could also use PackageKit (https://github.com/JuliaLang/BinDeps.jl/issues/66). `pkcon install -y` does not even require a password on my Fedora machine.

**Comment by Keno** [comment]
The problem is that detach no longer sets the controlling terminal. Blame @vtjnash ;)

**Comment by Keno** [comment]
See https://github.com/JuliaLang/julia/issues/17904

**Comment by vtjnash** [comment]
> This is a likely unintended side-effect of the fix for

FWIW, this was the *intended* effect. Our build slave should never become the controlling terminal master. Probably not a bad thing to prevent usage of `stdin` anyways, e.g. following the example of Debian (https://debian-handbook.info/browse/stable/sect.package-meta-information.html#sidebar.questions-conffiles#id-1.8.6.6.13.5).

**Comment by Keno** [comment]
Surely breaking BinDeps was not the intended effect. I also don't see what's wrong with passing the controlling terminal through to apt.

**Comment by SimonDanisch** [comment]
So, do I get this right, that the current fix is to install things manually?
This would make it difficult to give users a nice install experience.
Is fixing this on low priority?

**Comment by tkelman** [comment]
If someone can propose a fix we can try it. No one is actively working on this right now to my knowledge. I have plans to entirely replace bindeps in a way that doesn't rely on sudo or system package managers at all, but will certainly review any proposed changes here and try to keep it at least not getting any worse until a replacement is ready and widely usable (which is still a ways out).

**Comment by SimonDanisch** [comment]
So I was getting it to work by starting julia with sudo and passing -y to the apt-get command.... That'd be a bit better, but I heard people don't like either of these to be a default in a binary dependency installation...

**Comment by tkelman** [comment]
No, starting julia with sudo is a bad idea.

**Comment by SimonDanisch** [comment]
is there a way to figure out if we have the tty/askpass available, and if not replace the sudo command with an instruction?

**Comment by Keno** [comment]
The right solution is to pass through the controlling terminal as mentioned above.

**Comment by vtjnash** [comment]
> if we have the tty

never / not not supposed to be possible

> pass through the controlling terminal

no. we called `detach` specifically to prevent it from having a controlling terminal.

However, we could open a pseudo-tty pair (or use `sudo -S` and pass a normal pipe) and handle config IO ourselves.

**Comment by Keno** [comment]
> no. we called detach specifically to prevent it from having a controlling terminal.

We called detach before that stopped giving it a controlling terminal. What would be the purpose of doing this other than breaking people's installation of packages.

**Comment by vtjnash** [comment]
> I also don't see what's wrong with passing the controlling terminal through to apt.

The problem is you won't get it back. When `Pkg.build` finishes, it may kill all the other processes that were connected to that stdio (including the parent process and any other build processes).

**Comment by Keno** [comment]
> The problem is you won't get it back. When Pkg.build finishes, it may kill all the other processes that were connected to that stdio (including the parent process and any other build processes).

Are you sure that's not a process-group related problem rather than a controlling terminal problem?

**Comment by vtjnash** [comment]
Yes. They get SIGPIPE not SIGHUP.

**Comment by Keno** [comment]
Can you elaborate by what mechanism you think this happens?

**Comment by Keno** [comment]
Certainly apt can't kill all processes with the same controlling terminal, because otherwise you couldn't use it from bash.

**Comment by vtjnash** [comment]
> Can you elaborate by what mechanism you think this happens?

by watching it happen in the debugger, and wondering why the whole process was crashing down without printing any error messages (because it couldn't - stderr was closed).

**Comment by vtjnash** [comment]
> because otherwise you couldn't use it from bash

bash doesn't detach.

**Comment by Keno** [comment]
So what about running apt and then nicely asking for our controlling terminal back?

**Comment by SimonDanisch** [comment]
if it's polite enough, it might just work!

**Comment by Keno** [comment]
Just to be explicit about what I'm thinking here:
1. open /dev/tty
2. pass the controlling terminal to apt
3. wait until apt is done
4. ioctl(TIOCSCTTY) that fd

**Comment by vtjnash** [comment]
The `fd` may get shutdown when `apt` died. You have 3 and 4 reversed in that list. Also, you can't undo the effect of 1.

**Comment by Keno** [comment]
Know I mean `ioctl(TIOCSCTTY) that fd` back in julia when apt is done.

**Comment by vtjnash** [comment]
I'm confused then. Why are you talking about passing the controlling terminal to apt? You can't do that unless you spawn it with detach (it needs to be the session leader). The combination that you are suggesting doesn't make sense / isn't possible.

**Comment by Keno** [comment]
Yes, spawn it with detach, give up the controlling terminal and give it to apt, then take it back.

**Comment by vtjnash** [comment]
But then you can't take (or give) it back - the controlling terminal may already be shutdown.

**Comment by Keno** [comment]
How does one shut down a controlling terminal?

**Comment by vtjnash** [comment]
The kernel does it for you:

> The controlling terminal is inherited by a child process during a fork() function call. A process relinquishes its controlling terminal when it creates a new session with the setsid() function; other processes remaining in the old session that had this terminal as their controlling terminal continue to have it. **Upon the close of the last file descriptor in the system (whether or not it is in the current session) associated with the controlling terminal, it is unspecified whether all processes that had that terminal as their controlling terminal cease to have any controlling terminal. Whether and how a session leader can reacquire a controlling terminal after the controlling terminal has been relinquished in this fashion is unspecified.** A process does not relinquish its controlling terminal simply by closing all of its file descriptors associated with the controlling terminal if other processes continue to have it open.

> When a controlling process terminates, the controlling terminal is dissociated from the current session, allowing it to be acquired by a new session leader. **Subsequent access to the terminal by other processes in the earlier session may be denied, with attempts to access the terminal treated as if a modem disconnect had been sensed.**

IEEE Std 1003.1, 2004 http://pubs.opengroup.org/onlinepubs/009604599/basedefs/xbd_chap11.html

**Comment by Keno** [comment]
Yeah, we don't close all the file descriptors. Whether we're allowed to reacquire is unspecified according to POSIX, but is there any operating system that actually disallows this?

**Comment by Keno** [comment]
Actually I think the second paragraph says something different. It says that if a session leader dies, access to the terminal is unspecified, it doesn't forbid reacquisition by a different session leader.

**Comment by vtjnash** [comment]
you can argue with the lawyer-ees all you will, but https://github.com/JuliaLang/BinDeps.jl/issues/199#issuecomment-277082954

**Comment by Keno** [comment]
Ok, so your claim is that when a session leader exits, all file descriptors associated with that session leader's controlling terminal thereafter produce SIGPIPE?

**Comment by vtjnash** [comment]
Yes, that it what happened.

**Comment by vtjnash** [comment]
Also, the user loses the ability to interrupt or signal the host process (the REPL) via the keyboard as soon as the controlling terminal is lost.

**Comment by Keno** [comment]
```
$ cat testtty.c
#include <sys/types.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <termios.h>
#include <stdio.h>

int main() {
    if (vfork() == 0) {
        setsid();
        ioctl(0, TIOCSCTTY);
        _exit(0);
    }
    printf("I'm doing science and I'm still alive\n");
    return 0;
}
```

**Comment by Keno** [comment]
Sorry, made a mistake, hold on.

**Comment by Keno** [comment]
Actually, I'm confused. The terminal should already be the controlling terminal of some other process group. Why did it used to get stolen at all?

**Comment by vtjnash** [comment]
> Why did it used to get stolen at all?

It didn't, except for a couple of months on master. There was a libuv bug I accidentally didn't prevent from becoming a Julia bug when moving from v0.9 -> v1.7 (I was aware of it before then, which is why it wasn't a bug in our fork of v0.9, but mistakenly thought I could drop the revert commit during the rebase).

**Comment by Keno** [comment]
Ok, but this worked at some point after we started detaching, but before we stopped setting ctty, so somehow it got a controlling terminal.

**Comment by vtjnash** [comment]
Also, I'm not sure what you're end goal is here. That we should badly emulate the effect of not calling `detach`, while still calling `detach`?

**Comment by vtjnash** [comment]
> Ok, but this worked at some point after we started detaching, but before we stopped setting ctty, so somehow it got a controlling terminal.

Yes, that's when / how I noticed the libuv bug had been introduced to master.

**Comment by Keno** [comment]
The goal is for sudo to be able to get at the controlling terminal so it can ask the user for a password.

**Comment by Keno** [comment]
> Yes, that's when / how I noticed the libuv bug had been introduced to master.

I don't understand this mechanism though. Why did that terminal not already have a controlling terminal?

**Comment by vtjnash** [comment]
If that's the goal, then Pkg.build needs to stop being the daemon it was implemented to be (https://github.com/JuliaLang/julia/pull/13506). Daemon's aren't supposed to be the controlling leader, or they wouldn't be daemons.

**Comment by Keno** [comment]
I thought somebody had good reasons for putting detach there. If not, then yes, it should be removed.

**Comment by vtjnash** [comment]
> Why was that terminal not already associated with a process group?

I don't understand the question. It was associated with the process leader (Julia). The detached child process (libuv) was explicitly taking steps to ensure that it would steal the process leader away from its parent.

**Comment by Keno** [comment]
ctty stealing is not allowed unless you're root on linux.

**Comment by vtjnash** [comment]
While you're writing your fancy C programs, I would perhaps be remiss if I failed to point out that the original submitter of the bug report noting the breakage caused by libuv was a guy named loladiro (https://github.com/JuliaLang/julia/issues/16398). 😉

**Comment by Keno** [comment]
That's under lldb IIRC

**Comment by haberdashPI** [comment]
In reading through this thread again I find myself very confused. 

Is this affecting *all* binary dependencies on Linux? How is, for example, Gtk.jl, successfully installed via its build.jl script (it's clear that [it is](http://pkg.julialang.org/logs/Gtk_0.5.log))? I've tried to port my package to linux but have run into this problem. (Granted I am doing so via a VirtualBox install of Ubuntu, so maybe there is something specific to my environment that I'm running into...??).

**Comment by tkelman** [comment]
PkgEval gets around this by setting apt-get to always install things without bothering to ask for confirmation: https://github.com/JuliaCI/PackageEvaluator.jl/blob/60823fb6eebc69830368aa2a8ea6c84b26722695/scripts/setup.sh#L19-L27

**Comment by haberdashPI** [comment]
Got it! Thanks, that makes more sense.

**Comment by cstjean** [comment]
I'm getting this error [on travis](https://travis-ci.org/cstjean/TraceCalls.jl/jobs/272297769#L647-L654), but only on Linux. It [was working fine](https://travis-ci.org/cstjean/TraceCalls.jl/jobs/270668591#L442-L444) five days ago. I don't think I've made any change to my package that would affect this. Did anything else change?

**Comment by iblis17** [comment]
@cstjean I think your issue is unrelated.

[the passed build is on precise](https://travis-ci.org/cstjean/TraceCalls.jl/jobs/270668591#L9)

[the failed build is on trusty](https://travis-ci.org/cstjean/TraceCalls.jl/jobs/272297769#L9) which is the container-based runner. IIRC, `sudo` is not allowed on it.
